stake-pool/cli/src/main.rs

```diff
@@ -406,47 +406,59 @@ fn prepare_withdraw_accounts(
-    accounts.sort_by(|a, b| a.1.lamports.cmp(&b.1.lamports));
+    // Sort from highest to lowest balance
+    accounts.sort_by(|a, b| b.1.lamports.cmp(&a.1.lamports));
 
     // Prepare the list of accounts to withdraw from
     let mut withdraw_from: Vec<WithdrawAccount> = vec![];
     let mut remaining_amount = amount;
 
-    // First use largest amounts
-    while !accounts.is_empty() && accounts.last().unwrap().1.lamports < remaining_amount {
-        let (pubkey, account) = accounts.last().unwrap();
+    let mut split_candidate: Option<WithdrawAccount> = None;
 
-        // Those accounts will be withdrawn completely with `claim` instruction
-        withdraw_from.push(WithdrawAccount {
-            pubkey: *pubkey,
-            account: account.clone(),
-            amount: account.lamports,
-        });
-        remaining_amount -= account.lamports;
-        accounts.pop();
-    }
-    if remaining_amount > 0 {
-        // Add first amount larger or equal to the remaining amount
-        for (pubkey, account) in accounts {
-            if account.lamports < remaining_amount {
-                continue;
+    // If amount to withdraw is more than the largest account
+    // then use up all the large accounts first
+    // Then either find the smallest account with balance more than remaining
+    // or find account to claim with exactly the same balance as remaining
+    for (pubkey, account) in accounts {
+        if account.lamports <= remaining_amount {
+            if split_candidate.is_some() {
+                // If there is an active split candidate and current account is smaller than remaining
+                // then break to use split candidate
+                if account.lamports < remaining_amount {
+                    break;
+                }
+                // Otherwise (account balance == remaining) procede with `claim`
             }
-
-            // This account will be withdrawn partially (if remaining_amount < account.lamports), so we put it first
-            withdraw_from.insert(
-                0,
-                WithdrawAccount {
-                    pubkey,
-                    account,
-                    amount: remaining_amount,
-                },
-            );
-            remaining_amount = 0;
-            break;
+            // Those accounts will be withdrawn completely with `claim` instruction
+            withdraw_from.push(WithdrawAccount {
+                pubkey,
+                account: account.clone(),
+                amount: account.lamports,
+            });
+            remaining_amount -= account.lamports;
+            if remaining_amount == 0 {
+                // We filled all the balance, ignore split candidate
+                split_candidate = None;
+                break;
+            }
+        } else {
+            // Save the last account with balance more than remaining as a candidate for split
+            split_candidate = Some(WithdrawAccount {
```

Comment:
For your example of "Withdraw 19 returns [10, 8, 1 of 2]", wouldn't `split_candidate` get overwritten on the second iteration by the account with 8?  Does the logic from 433-443 need to get done here too?

Actually, could you add some tests for all the cases you've laid out?  Then we'll be sure that everything works as intended

Reply:
No, it is not, iterations will go like this:
1. Add 10 to the list, remaining is 9
2. Add 8 to the list, remaining is 8
3. Split candidate set to 7, continue
4. Split candidate is 5
5. Split candidate is 3
6. Spit candidate is 2
Now the cycle is over so we take 1 from the last split candidate.

Reply:
But you're right, I also thought we should cover it with tests at least for those cases I specified, will do.

Reply:
Yes, you're right, my mistake!